File: labs/python/example_portokasse_link_check.py
"""
Portokasse linkage diagnostic (no stamp order, no charge).

Checks DHL app token + Portokasse user auth via Internetmarke REST v1.
"""

from __future__ import annotations

import asyncio
import json
import os
import sys
from datetime import UTC, datetime
from pathlib import Path
from typing import Any

_REPO_ROOT = Path(__file__).resolve().parents[2]
if str(_REPO_ROOT) not in sys.path:
    sys.path.insert(0, str(_REPO_ROOT))

from labs.lib.python.internetmarke_auth_diagnostic import (  # noqa: E402
    exit_code_for_auth_status,
    probe_internetmarke_auth,
)
from labs.lib.python.load_env import load_lab_env  # noqa: E402

load_lab_env()

from porto_sdk.adapters.deutschepost.internetmarke.bootstrap import (  # noqa: E402
    get_internetmarke_base_url,
    load_internetmarke_config,
)


def mask(value: str | None) -> str:
    if not value:
        return "MISSING"
    if len(value) <= 6:
        return "***"
    return f"{value[:3]}...{value[-3:]}"


def save_result(payload: dict[str, Any]) -> Path:
    observer_run_dir = os.getenv("OBSERVER_RUN_DIR")
    if observer_run_dir:
        out_dir = Path(observer_run_dir) / "auth"
    else:
        out_dir = Path(__file__).parent / "artifacts" / "auth"
    out_dir.mkdir(parents=True, exist_ok=True)

    ts = datetime.now(UTC).strftime("%Y%m%d-%H%M%S")
    out_path = out_dir / f"portokasse_link_check_{ts}.json"
    out_path.write_text(json.dumps(payload, indent=2) + "\n", encoding="utf-8")
    return out_path


async def main() -> int:
    im = load_internetmarke_config("deutschepost")
    creds = im.credentials if im else {}
    base_url = get_internetmarke_base_url(im)

    print("Portokasse linkage diagnostic")
    print("=" * 34)
    print(f"DHL API key:    {mask(creds.get('dhl_api_key'))}")
    print(f"DHL API secret: {mask(creds.get('dhl_api_secret'))}")
    print(f"Base URL:       {base_url}")
    print(f"Portokasse user:{mask(creds.get('username'))}")

    auth = await probe_internetmarke_auth(
        base_url=base_url,
        username=creds.get("username"),
        password=creds.get("password"),
        api_key=creds.get("dhl_api_key"),
        api_secret=creds.get("dhl_api_secret"),
    )

    path = save_result(
        {
            "ts": datetime.now(UTC).isoformat(),
            **auth.to_dict(),
            "checks": {
                "app_auth": {
                    "endpoint": f"{base_url.rstrip('/')}/auth/token",
                    "status_code": auth.app_status,
                    "body_preview": auth.app_body_preview,
                },
                "user_auth": {
                    "endpoint": f"{base_url.rstrip('/')}/user/authenticate",
                    "status_code": auth.user_status,
                    "body_preview": auth.user_body_preview,
                },
            },
        }
    )

    print(f"\nstatus: {auth.status}")
    print(f"blocking_stage: {auth.blocking_stage}")
    print(f"hint: {auth.hint}")
    if auth.next_steps:
        print("next_steps:")
        for step in auth.next_steps:
            print(f"  - {step}")
    print(f"saved: {path}")
    return exit_code_for_auth_status(auth.status)


if __name__ == "__main__":
    raise SystemExit(asyncio.run(main()))

File: labs/typescript/example_portokasse_link_check.ts
/**
 * Portokasse linkage diagnostic (TypeScript) — no stamp order, no charge.
 */

import { mkdirSync, writeFileSync } from "node:fs";
import { dirname, join } from "node:path";
import { fileURLToPath } from "node:url";
import {
  getInternetmarkeBaseUrl,
  loadInternetmarkeConfig,
} from "../../../sdks/porto-sdk-typescript/src/index.ts";
import {
  exitCodeForAuthStatus,
  probeInternetmarkeAuth,
} from "../../lib/typescript/internetmarke_auth_diagnostic.ts";
import { loadLabEnv } from "../../lib/typescript/load_env.ts";

const __dirname = dirname(fileURLToPath(import.meta.url));
loadLabEnv();

function mask(value: string | undefined): string {
  if (!value) return "MISSING";
  if (value.length <= 6) return "***";
  return `${value.slice(0, 3)}...${value.slice(-3)}`;
}

function saveResult(payload: Record<string, unknown>): string {
  const outDir = process.env.OBSERVER_RUN_DIR
    ? join(process.env.OBSERVER_RUN_DIR, "auth")
    : join(__dirname, "artifacts", "auth");
  mkdirSync(outDir, { recursive: true });
  const ts = new Date().toISOString().replace(/[:.]/g, "-").slice(0, 19);
  const outPath = join(outDir, `portokasse_link_check_${ts}.json`);
  writeFileSync(outPath, `${JSON.stringify(payload, null, 2)}\n`, "utf8");
  return outPath;
}

async function main(): Promise<number> {
  const im = loadInternetmarkeConfig("deutschepost");
  const creds = im?.credentials ?? {};
  const baseUrl = getInternetmarkeBaseUrl(im);

  console.log("Portokasse linkage diagnostic (TS)");
  console.log(`DHL API key: ${mask(creds.dhl_api_key)}`);
  console.log(`Portokasse user: ${mask(creds.username)}`);

  const auth = await probeInternetmarkeAuth({
    baseUrl,
    username: creds.username,
    password: creds.password,
    apiKey: creds.dhl_api_key,
    apiSecret: creds.dhl_api_secret,
  });

  const path = saveResult({
    ts: new Date().toISOString(),
    ...auth,
    checks: {
      app_auth: {
        endpoint: `${baseUrl.replace(/\/$/, "")}/auth/token`,
        status_code: auth.app_status,
        body_preview: auth.app_body_preview,
      },
      user_auth: {
        endpoint: `${baseUrl.replace(/\/$/, "")}/user/authenticate`,
        status_code: auth.user_status,
        body_preview: auth.user_body_preview,
      },
    },
  });

  console.log(`status: ${auth.status}`);
  console.log(`blocking_stage: ${auth.blocking_stage}`);
  console.log(`hint: ${auth.hint}`);
  if (auth.next_steps.length > 0) {
    console.log("next_steps:");
    for (const step of auth.next_steps) {
      console.log(`  - ${step}`);
    }
  }
  console.log(`saved: ${path}`);
  return exitCodeForAuthStatus(auth.status);
}

main().then((code) => process.exit(code));
